### ISSUE TITLE
[css-scroll-snap] Difficult to use snap areas which are larger than the scrollport

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#6863]
The css-scroll-snap spec makes it possible for [snap areas larger than the scrollport](https://drafts.csswg.org/css-scroll-snap/#snap-overflow) to be accessed, however, the behavior in many browsers makes it quite difficult and results in unexpected scrolls. For example, If you test the examples on https://output.jsbin.com/pojeyeg it is difficult to reach the end of a snap area without snapping to the next. Also, it is difficult to get to the previous element in the scroll port unless you scroll very far such that the top / left of the element is closer than the current one. In a [more extreme example](https://jsbin.com/mivococ/edit?html,css,output) it is very difficult to scroll to the previous item unless you fling.

I think we should update the spec to consider the entire area of valid snap positions (per [css-scroll-snap section 5.2.2](https://drafts.csswg.org/css-scroll-snap/#snap-overflow)) when determining which snap point to snap to, and only snap as far as necessary to reach that next valid snap point rather than returning to the start of the oversized snap area.

@kevers-google do you have additional thoughts about what the spec text should look like.

**Comment by flackr** [comment]
As some additional examples of what this looks like, I've recorded a couple videos of this in chrome and safari:

## Chrome
![chrome](https://user-images.githubusercontent.com/366761/144922433-8baae178-d0fa-4f18-aba7-d4b12edb5660.gif)

## Safari
![safari](https://user-images.githubusercontent.com/366761/144922474-3927e90c-fb96-44d5-88a6-e848b2454b98.gif)

Firefox has similar behavior.

**Comment by fantasai** [comment]
What you're asking for is definitely what the spec intended in 5.2.2, so that seems like a UA bug?

I see that the wording added in https://github.com/w3c/csswg-drafts/commit/6964742efa036bbe0b7e0f5804d63b301262829e made the key sentence confusing, though.

**Comment by fantasai** [comment]
Expanded the explanation of the example a bit and split up the nearly-impossible-to-understand sentence into pieces so it can be grammatically parsed by mere humans... so hopefully the spec should be a bit more understandable on this point now.

**Comment by tabatkins** [comment]
Yeah, we realized when discussing this that the current spec definition isn't correct, but we've had a lot of trouble figuring out what it *should* be.

The constraints we're trying to apply are:

1. *Every* part of a snap area *must* be visible from some snap position. This is strictly required.
2. We want to add the minimum number of additional snap positions to satisfy 1.
3. We don't want to be "arbitrary"; like, don't add snap points ever 20px or something, because the user didn't ask for that.

Figuring out how to turn those constraints into straightforward rules has turned out to be far more complex than expected.

**Comment by hiikezoe** [comment]
A confusing point was (at least for me) this [the snap area covers the snapport](https://drafts.csswg.org/css-scroll-snap/#snap-overflow) condition. For example, given that the snapport size is 100px, and the snap area size is 150px, but the top 100px of the snap area is scrolled out, then it does cover the snapport? I thought it doesn't. It does?

**Comment by hiikezoe** [comment]
Also note that in cases of keyboard scrolling a sentence in Choosing Snap Positions, [This implies that a scroll with only an intended direction must always ignore the starting snap positions](https://drafts.csswg.org/css-scroll-snap/#choosing), will be an obstacle to make the expected behavior happen. That means a test added in https://github.com/web-platform-tests/wpt/pull/31930 doesn't work as expected due to the intended direction's restriction.

**Comment by zcorpan** [comment]
This issue is blocking https://bugzilla.mozilla.org/show_bug.cgi?id=1744876 which is part of the Scrolling focus area in Interop 2023.
